### PR TITLE
feat/benchmark convolutional network

### LINKED ISSUE #126: Add benchmark
## Feature Description
Add a simple convolutional network in the tests where we do a feedforward using ```VirtualMachines```.

## Is your feature request related to a problem?
 - Transparency, if someone wants to check the times for doing a forward pass.
 - Have a clear vision on how we impact the running time

### PR DESCRIPTION
**Pull request by danielorihuela**
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

## Explanation

The workflow added runs pytest-benchmark for all the tests that can be found in `tests/benchmarks`. The result is then pushed to `gh-pages` branch, to `dev/bench` folder. The branch and folder name can be changed https://github.com/rhysd/github-action-benchmark#action-inputs.

Closes #126

### CODE REVIEW

File: tests/benchmarks/module/module_benchmark_test.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    res_mpc = mpc_model(x_mpc)
+    assert isinstance(res_mpc, MPCTensor)
+
+    res = res_mpc.reconstruct()
+    expected = expected.detach().numpy()
+    assert np.allclose(res, expected, atol=1e-3)
+
+
+def test_run_conv_model(benchmark, get_clients: Callable[[int], List[Any]]):
```
Comment (gmuraru) [inline tests/benchmarks/module/module_benchmark_test.py:61]: Q: Isn't this going to run every time we run the tests?
Ideally, we would like this to be run every time we do a "merge" into the ```main``` branch.

Comment (danielorihuela) [reply tests/benchmarks/module/module_benchmark_test.py:61]: Yes, depending on the command you use. `pytest` or `pytests tests` will execute them all. But you could do `pytest --deselect=tests/benchmarks.

Also, this is going to run every time a "push" is done to main, thanks to [benchmarks.yml](https://github.com/OpenMined/SyMPC/pull/128/commits/be7efbe7ff59f216b72b9af931a9eeef76675263#diff-22eef62c9d7b36d02f5fd5aaada92702e22d6795c464239cb7b8fe0f26ea1e1c).

I can give it another though, if you want. And see what can be done to not execute this folder when running `pytest`, or move it to a different folder named `benchmarks` next to `tests`.

**Review by aanurraj — CHANGES_REQUESTED** [review]

File: .github/workflows/benchmarks.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  push:
+    branches:
+      - main
+    paths:
+      - "**.py"
+      - "**.ipynb"
+      - "setup.cfg"
+      - ".github/workflows/**.yml"
+      - "requirements_dev.txt"
+      - "requirements.txt"
```
[the remark this answers is not in the document]
Comment (danielorihuela) [reply .github/workflows/benchmarks.yml:13]: Hmmm, its an error. When I was basing this new action in the pysyft benchmark and sympc tests I forgot to delete those. We want to delete them, otherwise, the benchmarks will be updated even when the changes are not being pushed to main.
If am I right, please, tell it to me and I will change it right now.

Thanks for the feedback!

[the remark this answers is not in the document]
Comment (danielorihuela) [reply .github/workflows/benchmarks.yml]: @aanurraj I have already done the change. Whenever you want, feel free to resolve the issue if I have done it correctly. Thanks!

File: .github/workflows/benchmarks.yml [hunk 3]
```diff
@@ -4,13 +4,6 @@ on:
   push:
     branches:
       - main
-    paths:
```
Comment (aanurraj) [inline .github/workflows/benchmarks.yml]: you should include 
```
- "**.py"
- ".github/workflows/**.yml"
```

in the workflow so that if any changes are made in these file the action will be triggered.

File: .github/workflows/benchmarks.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (aanurraj) [reply .github/workflows/benchmarks.yml]: I made a comment below @danielorihuela

File: .github/workflows/benchmarks.yml (hunk 3 above)
Comment (danielorihuela) [reply .github/workflows/benchmarks.yml]: I do not think we should put it. In that case, if I do a new MR where I change the benchmarks.yml a new push with new benchmarks will be done. Or at least is what I understand.

Comment (aanurraj) [reply .github/workflows/benchmarks.yml]: `pull_request: `--> this is for PR
`push:` -->  is for merge

you have to add above paths to `push:`

if it makes sense ?

File: tests/benchmarks/module/module_benchmark_test.py (hunk 1 above)
Comment (aanurraj) [inline tests/benchmarks/module/module_benchmark_test.py:61]: **Suggestion:** can you move this function to separate file ? may be in future we will be adding more benchmarks to it.

**Review by aanurraj — CHANGES_REQUESTED** [review]

File: .github/workflows/benchmarks.yml (hunk 3 above)
Comment (danielorihuela) [reply .github/workflows/benchmarks.yml]: Aaaaaaaah, okey. The action is triggered when a "merge" is made to "main" and it changes one of the following "paths". If what I said is correct, it makes sense.

Comment (aanurraj) [reply .github/workflows/benchmarks.yml]: yess... 😃
no problem

File: tests/benchmarks/module/module_benchmark_test.py (hunk 1 above)
Comment (danielorihuela) [reply tests/benchmarks/module/module_benchmark_test.py:61]: The run_conv_model, sure. I will try it later this afternoon.

Comment (gmuraru) [reply tests/benchmarks/module/module_benchmark_test.py:61]: I think it is ok for the moment for having them in the tests

**Review by gmuraru — APPROVED** [review]
LGTM!

**Review by aanurraj — APPROVED** [review]

File: tests/benchmarks/module/module_benchmark_test.py (hunk 1 above)
Comment (danielorihuela) [reply tests/benchmarks/module/module_benchmark_test.py:9]: Done, for the moment I decided to do a single file called `conv_model`. In the future we could group different models by file: conv_models, linear_models, ... or we could use one file for each. In any case, changing the name in the future should be easy. If you are okey with the change, we can merge it!

Comment (danielorihuela) [reply tests/benchmarks/module/module_benchmark_test.py]: While doing aanuraj suggestion and fixing the errors, I decided to move benchmarks into another folder different from the tests folder to tackle gmuraru first question.
@aanurraj @gmuraru If you could do a last review before merging, would be awesome. Thanks!

**Comment by gmuraru** [comment]
Great Work!
